File: app/src/main/java/com/fatbook/fatbookapp/ui/fragment/FeedFragment.java
package com.fatbook.fatbookapp.ui.fragment;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;

import com.fatbook.fatbookapp.R;
import com.fatbook.fatbookapp.core.Ingredient;
import com.fatbook.fatbookapp.core.Recipe;
import com.fatbook.fatbookapp.core.User;
import com.fatbook.fatbookapp.databinding.FragmentFeedBinding;
import com.fatbook.fatbookapp.retrofit.RetrofitFactory;
import com.fatbook.fatbookapp.ui.activity.SplashActivity;
import com.fatbook.fatbookapp.ui.adapters.RecipeAdapter;
import com.fatbook.fatbookapp.ui.listeners.OnRecipeClickListener;
import com.fatbook.fatbookapp.ui.listeners.OnRecipeRevertDeleteListener;
import com.fatbook.fatbookapp.ui.viewmodel.RecipeViewModel;
import com.fatbook.fatbookapp.ui.viewmodel.UserViewModel;
import com.fatbook.fatbookapp.util.KeyboardActionUtil;
import com.fatbook.fatbookapp.util.UserUtils;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

import lombok.extern.java.Log;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Log
public class FeedFragment extends Fragment implements OnRecipeClickListener, OnRecipeRevertDeleteListener {

    private FragmentFeedBinding binding;

    private List<Recipe> feedRecipeList;

    private RecipeViewModel recipeViewModel;

    private UserViewModel userViewModel;

    private User user;

    private RecipeAdapter adapter;

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requireActivity().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);

        recipeViewModel = new ViewModelProvider(requireActivity()).get(RecipeViewModel.class);
        userViewModel = new ViewModelProvider(requireActivity()).get(UserViewModel.class);

        if (recipeViewModel.getSelectedRecipePosition().getValue() != null) {
            if (user != null) {
                loadUser(recipeViewModel.getSelectedRecipePosition().getValue());
            }
        }

        feedRecipeList = new ArrayList<>();

        userViewModel.getUser().observe(getViewLifecycleOwner(), userUpdated -> {
            user = userUpdated;
            if (userViewModel.getFeedRecipeList().getValue() == null) {
                loadData();
            }
        });

        binding.swipeRefreshBookmarks.setColorSchemeColors(getResources().getColor(R.color.color_pink_a200));

        binding.swipeRefreshBookmarks.setOnRefreshListener(this::loadData);

        userViewModel.getFeedRecipeList().observe(getViewLifecycleOwner(), recipeList -> {
            feedRecipeList = recipeList;
            if (feedRecipeList == null) {
                feedRecipeList = new ArrayList<>();
            }
            adapter.setData(recipeList, user);
            adapter.notifyDataSetChanged();
        });
        setupAdapter();
        feedRecipeList = userViewModel.getFeedRecipeList().getValue();
        setupSearch();
    }

    private void setupSearch() {
        binding.editTextFeedSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                filter(editable.toString());
            }
        });
    }

    private void filter(String text) {
        try {
            List<Recipe> temp = new ArrayList<>();
            for (Recipe r : feedRecipeList) {
                if (StringUtils.containsIgnoreCase(r.getName(), text) || StringUtils.containsIgnoreCase(r.getDescription(), text)) {
                    temp.add(r);
                }
            }
            adapter.updateList(temp);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void setupAdapter() {
        RecyclerView recyclerView = binding.rvFeed;
        adapter = new RecipeAdapter(binding.getRoot().getContext(), feedRecipeList, user, this);
        ((SimpleItemAnimator) recyclerView.getItemAnimator()).setSupportsChangeAnimations(false);
        recyclerView.setAdapter(adapter);
    }

    //TODO убрать костыль в запросе
    private void loadData() {
        RetrofitFactory.apiServiceClient().getFeed(0L).enqueue(new Callback<List<Recipe>>() {
            @Override
            public void onResponse(@NonNull Call<List<Recipe>> call, @NonNull Response<List<Recipe>> response) {
                userViewModel.setFeedRecipeList(response.body());
                log.log(Level.INFO, "feed data load: SUCCESS");
                if (response.body() != null) {
                    log.log(Level.INFO, "loaded " + response.body().size() + " recipes");
                }
                loadUser(null);
            }

            @Override
            public void onFailure(@NonNull Call<List<Recipe>> call, @NonNull Throwable t) {
                userViewModel.setFeedRecipeList(new ArrayList<>());
                log.log(Level.INFO, "feed data load: FAILED");
            }
        });
    }

    @Override
    public void onRecipeClick(int position) {
        Recipe recipe = feedRecipeList.get(position);
        recipeViewModel.setSelectedRecipe(recipe);
        recipeViewModel.setSelectedRecipePosition(position);
        NavHostFragment.findNavController(this).navigate(R.id.action_go_to_recipe_view);
    }

    @Override
    public void onBookmarksClick(Recipe recipe, boolean bookmark, int position) {
        recipeBookmarked(recipe, bookmark, position);
    }

    private void recipeBookmarked(Recipe recipe, boolean bookmark, int position) {
        RetrofitFactory.apiServiceClient().recipeBookmarked(user.getPid(), recipe.getPid(), bookmark).enqueue(new Callback<Recipe>() {
            @Override
            public void onResponse(@NonNull Call<Recipe> call, @NonNull Response<Recipe> response) {
                log.log(Level.INFO, "bookmark SUCCESS");
                recipeViewModel.setSelectedRecipe(response.body());
                loadUser(position);
            }

            @Override
            public void onFailure(@NonNull Call<Recipe> call, @NonNull Throwable t) {
                log.log(Level.INFO, "bookmark FAILED");
            }
        });
    }

    @Override
    public void onForkClicked(Recipe recipe, boolean fork, int position) {
        recipeForked(recipe, fork, position);
    }

    private void recipeForked(Recipe recipe, boolean fork, int position) {
        RetrofitFactory.apiServiceClient().recipeForked(user.getPid(), recipe.getPid(), fork).enqueue(new Callback<Recipe>() {
            @Override
            public void onResponse(@NonNull Call<Recipe> call, @NonNull Response<Recipe> response) {
                log.log(Level.INFO, "fork SUCCESS");
                recipeViewModel.setSelectedRecipe(response.body());
                loadUser(position);
            }

            @Override
            public void onFailure(@NonNull Call<Recipe> call, @NonNull Throwable t) {
                log.log(Level.INFO, "fork FAILED");
            }
        });
    }

    private void loadUser(Integer position) {
        if (user == null) {
            SharedPreferences sharedPreferences = requireActivity().getSharedPreferences(UserUtils.APP_PREFS, Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString(UserUtils.USER_LOGIN, StringUtils.EMPTY);
            editor.apply();
            startActivity(new Intent(requireActivity(), SplashActivity.class));
            requireActivity().finish();
        } else {
            RetrofitFactory.apiServiceClient().getUser(user.getLogin()).enqueue(new Callback<User>() {
                @Override
                public void onResponse(@NonNull Call<User> call, @NonNull Response<User> response) {
                    if (response.code() == 200) {
                        log.log(Level.INFO, "user load SUCCESS");
                        userViewModel.setUser(response.body());
                        System.out.println(user);
                        adapter.setUser(user);
                        if (position != null) {
                            if (recipeViewModel.getSelectedRecipe().getValue() != null) {
                                feedRecipeList.get(position).setForks(recipeViewModel.getSelectedRecipe().getValue().getForks());
                                adapter.notifyItemChanged(position);
                            } else {
                                feedRecipeList.remove((int) position);
                                adapter.notifyItemRemoved(position);
                            }
                            recipeViewModel.setSelectedRecipe(null);
                            recipeViewModel.setSelectedRecipePosition(null);
                        } else {
                            adapter.notifyDataSetChanged();
                        }
                    } else {
                        log.log(Level.INFO, "user load FAILED " + response.code());
                    }
                    if (binding != null) {
                        binding.swipeRefreshBookmarks.setRefreshing(false);
                    }
                }

                @Override
                public void onFailure(@NonNull Call<User> call, @NonNull Throwable t) {
                    log.log(Level.INFO, "user load FAILED");
                }
            });
        }
    }

    @Override
    public void onRecipeRevertDeleteClick(Recipe recipe) {
        //TODO revert delete recipe
        /**
         * post recipe again
         */
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentFeedBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onResume() {
        super.onResume();
        binding.getRoot().getViewTreeObserver().addOnGlobalLayoutListener(new KeyboardActionUtil(binding.getRoot(), requireActivity()).listenerForAdjustResize);
    }

    @Override
    public void onPause() {
        super.onPause();
        binding.getRoot().getViewTreeObserver().removeOnGlobalLayoutListener(new KeyboardActionUtil(binding.getRoot(), requireActivity()).listenerForAdjustResize);
    }

}

File: app/src/main/java/online/fatbook/fatbookapp/ui/fragment/FeedFragment.kt
package online.fatbook.fatbookapp.ui.fragment

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.NavHostFragment
import androidx.recyclerview.widget.SimpleItemAnimator
import kotlinx.android.synthetic.main.fragment_feed.*
import okhttp3.MultipartBody
import okhttp3.RequestBody
import online.fatbook.fatbookapp.R
import online.fatbook.fatbookapp.callback.ResultCallback
import online.fatbook.fatbookapp.core.authentication.LoginResponse
import online.fatbook.fatbookapp.core.recipe.Recipe
import online.fatbook.fatbookapp.core.user.User
import online.fatbook.fatbookapp.databinding.FragmentFeedBinding
import online.fatbook.fatbookapp.retrofit.RetrofitFactory
import online.fatbook.fatbookapp.ui.activity.SplashActivity
import online.fatbook.fatbookapp.ui.adapters.RecipeAdapter
import online.fatbook.fatbookapp.ui.listeners.OnRecipeClickListener
import online.fatbook.fatbookapp.ui.listeners.OnRecipeRevertDeleteListener
import online.fatbook.fatbookapp.ui.viewmodel.AuthenticationViewModel
import online.fatbook.fatbookapp.ui.viewmodel.RecipeViewModel
import online.fatbook.fatbookapp.ui.viewmodel.UserViewModel
import online.fatbook.fatbookapp.util.KeyboardActionUtil
import online.fatbook.fatbookapp.util.UserUtils
import online.fatbook.fatbookapp.util.obtainViewModel
import org.apache.commons.lang3.StringUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class FeedFragment : Fragment(), OnRecipeClickListener, OnRecipeRevertDeleteListener {

    private var binding: FragmentFeedBinding? = null
    private var adapter: RecipeAdapter? = null

    private val authViewModel by lazy { obtainViewModel(AuthenticationViewModel::class.java) }
    private val recipeViewModel by lazy { obtainViewModel(RecipeViewModel::class.java) }
    private val userViewModel by lazy { obtainViewModel(UserViewModel::class.java) }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentFeedBinding.inflate(inflater, container, false)
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupSwipeRefresh()
        setupMenu()
        if (userViewModel.user.value == null) {
            login()
        } else {
            loadUserInfo()
        }

//        requireActivity().window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
//        if (recipeViewModel.selectedRecipePosition.value != null) {
//            if (userViewModel.user.value != null) {
//                loadUser(recipeViewModel.selectedRecipePosition.value)
//            }
//        }
//        setupAdapter()
//        addObservers()
//        loadUser(
//            requireActivity().getSharedPreferences(UserUtils.APP_PREFS, Context.MODE_PRIVATE)
//                .getString(UserUtils.USER_LOGIN, StringUtils.EMPTY)!!, null
//        )
    }

    //    ==========================================================================================

    private fun login() {
        progressbar_feed.visibility = View.VISIBLE
        val request: RequestBody = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("username", authViewModel.username.value!!)
            .addFormDataPart("password", authViewModel.password.value!!).build()
        authViewModel.login(request, object : ResultCallback<LoginResponse> {
            override fun onResult(value: LoginResponse?) {
                saveTokens(value!!)
                println("User ${authViewModel.username.value} logged in")
            }

            override fun onFailure(value: LoginResponse?) {
                login()
            }
        })
    }

    private fun saveTokens(value: LoginResponse) {
        authViewModel.jwtAccess.value = value.access_token
        authViewModel.jwtRefresh.value = value.refresh_token
        RetrofitFactory.updateJWT(value.access_token!!)
        loadUserInfo()
    }

    private fun loadUserInfo() {
        userViewModel.getUserByUsername(
            authViewModel.username.value!!,
            object : ResultCallback<User> {
                override fun onResult(value: User?) {
                    userViewModel.user.value = value
                    progressbar_feed.visibility = View.GONE
                }

                override fun onFailure(value: User?) {
                    loadUserInfo()
                }
            })
    }

    private fun setupSwipeRefresh() {
        swipe_refresh_bookmarks.setColorSchemeColors(
            ContextCompat.getColor(
                requireContext(),
                R.color.color_pink_a200
            )
        )
        swipe_refresh_bookmarks.setOnRefreshListener {
//            loadData()
        }
    }

    private fun setupMenu() {
        val activity = (activity as AppCompatActivity?)!!
        activity.setSupportActionBar(toolbar_feed)
        setHasOptionsMenu(true)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.feed_overflow_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.item_direct_messages -> {
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_go_to_direct_messages_from_feed)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun addObservers() {
        userViewModel.user.observe(viewLifecycleOwner) {
            android.util.Log.d("TAG", " test ")
        }
        userViewModel.feedRecipeList.observe(viewLifecycleOwner) {
            android.util.Log.d("TAG", " FEED ")
            adapter!!.setData(it ?: ArrayList(), userViewModel.user.value)
        }
    }

    private fun filter(text: String) {
        try {
            val temp: ArrayList<Recipe> = ArrayList()
            for (r in userViewModel.feedRecipeList.value!!) {
                if (StringUtils.containsIgnoreCase(
                        r.title, text
                    )
                ) {
                    temp.add(r)
                }
            }
            adapter!!.setData(temp)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun setupAdapter() {
        val recyclerView = binding!!.rvFeed
        adapter = RecipeAdapter()
        adapter!!.setData(ArrayList(), userViewModel.user.value)
        adapter!!.setClickListener(this)
        (recyclerView.itemAnimator as SimpleItemAnimator?)!!.supportsChangeAnimations = false
        recyclerView.adapter = adapter
    }

    //TODO убрать костыль в запросе
    private fun loadData() {
        RetrofitFactory.apiServiceClient().getFeed(0L)
            .enqueue(object : Callback<ArrayList<Recipe>?> {
                override fun onResponse(
                    call: Call<ArrayList<Recipe>?>, response: Response<ArrayList<Recipe>?>
                ) {
                    if (response.body() != null) {
                        userViewModel.feedRecipeList.value = response.body()
                    }
                    swipe_refresh_bookmarks.isRefreshing = false
//                    loadUser(null)
                }

                override fun onFailure(call: Call<ArrayList<Recipe>?>, t: Throwable) {
                    userViewModel.feedRecipeList.value = ArrayList()
                }
            })
    }

    override fun onRecipeClick(position: Int) {
        val recipe = userViewModel.feedRecipeList.value!![position]
        recipeViewModel.selectedRecipe.value = recipe
        recipeViewModel.selectedRecipePosition.value = position
        NavHostFragment.findNavController(this).navigate(R.id.action_go_to_recipe_view_from_feed)
    }

    override fun onBookmarksClick(recipe: Recipe?, bookmark: Boolean, position: Int) {
        recipeBookmarked(recipe, bookmark, position)
    }

    private fun recipeBookmarked(recipe: Recipe?, bookmark: Boolean, position: Int) {
        RetrofitFactory.apiServiceClient()
            .recipeBookmarked(userViewModel.user.value!!.pid, recipe!!.pid, bookmark)
            .enqueue(object : Callback<Recipe?> {
                override fun onResponse(call: Call<Recipe?>, response: Response<Recipe?>) {
//                    FeedFragment.log.log(Level.INFO, "bookmark SUCCESS")
                    recipeViewModel.selectedRecipe.value = response.body()
                    loadUser(userViewModel.user.value!!.username!!, position)
                }

                override fun onFailure(call: Call<Recipe?>, t: Throwable) {
//                    FeedFragment.log.log(Level.INFO, "bookmark FAILED")
                }
            })
    }

    override fun onForkClicked(recipe: Recipe?, fork: Boolean, position: Int) {
        recipeForked(recipe, fork, position)
    }

    private fun recipeForked(recipe: Recipe?, fork: Boolean, position: Int) {
        RetrofitFactory.apiServiceClient()
            .recipeForked(userViewModel.user.value!!.pid, recipe!!.pid, fork)
            .enqueue(object : Callback<Recipe?> {
                override fun onResponse(call: Call<Recipe?>, response: Response<Recipe?>) {
//                    FeedFragment.log.log(Level.INFO, "fork SUCCESS")
                    recipeViewModel.selectedRecipe.value = response.body()
                    loadUser(userViewModel.user.value!!.username!!, position)
                }

                override fun onFailure(call: Call<Recipe?>, t: Throwable) {
//                    FeedFragment.log.log(Level.INFO, "fork FAILED")
                }
            })
    }

    private fun loadUser(login: String, position: Int?) {
//        if (userViewModel.user.value == null) {
        if (login == null) {
            val sharedPreferences =
                requireActivity().getSharedPreferences(UserUtils.APP_PREFS, Context.MODE_PRIVATE)
            val editor = sharedPreferences.edit()
            editor.putString(UserUtils.USER_LOGIN, StringUtils.EMPTY)
            editor.apply()
            startActivity(Intent(requireActivity(), SplashActivity::class.java))
            requireActivity().finish()
        } else {
            RetrofitFactory.apiServiceClient().getUserByUsername(login)
                .enqueue(object : Callback<User?> {
                    override fun onResponse(call: Call<User?>, response: Response<User?>) {
                        if (response.code() == 200) {
//                            FeedFragment.log.log(Level.INFO, "user load SUCCESS")
                            userViewModel.user.value = response.body()
                            println(userViewModel.user.value)
                            adapter!!.setUser(userViewModel.user.value!!)
//                            if (position != null) {
//                                if (recipeViewModel.selectedRecipe.value != null) {
//                                    userViewModel.feedRecipeList.value!![position].forks =
//                                        recipeViewModel.selectedRecipe.value!!.forks
//                                    adapter!!.notifyItemChanged(position)
//                                } else {
//                                    userViewModel.feedRecipeList.value!!.removeAt(position)
//                                    adapter!!.notifyItemRemoved(position)
//                                }
//                                recipeViewModel.selectedRecipe.value = null
//                                recipeViewModel.selectedRecipePosition.value = null
//                            } else {
//                                adapter!!.notifyDataSetChanged()
//                            }
                        } else {
//                            FeedFragment.log.log(Level.INFO, "user load FAILED " + response.code())
                        }
                        if (binding != null) {
                            binding!!.swipeRefreshBookmarks.isRefreshing = false
                        }
                    }

                    override fun onFailure(call: Call<User?>, t: Throwable) {
//                        FeedFragment.log.log(Level.INFO, "user load FAILED")
                    }
                })
        }
    }

    override fun onRecipeRevertDeleteClick(recipe: Recipe?) {
        //TODO revert delete recipe
        /**
         * post recipe again
         */
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onResume() {
        super.onResume()
        binding!!.root.viewTreeObserver.addOnGlobalLayoutListener(
            KeyboardActionUtil(
                binding!!.root, requireActivity()
            ).listenerForAdjustResize
        )
    }

    override fun onPause() {
        super.onPause()
        binding!!.root.viewTreeObserver.removeOnGlobalLayoutListener(
            KeyboardActionUtil(
                binding!!.root, requireActivity()
            ).listenerForAdjustResize
        )
    }
}